keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -893,6 +894,48 @@ of the IP family of the service. A single service which is dual-stack may use
 the same NodePort on both families.  `kube-proxy` will ensure that traffic is
 routed according to the families assigned for services.
 
+
+##### Spec.loadBalancerIP
+
+The `spec.loadBalancerIP` is a string and only allow one load-balancer
+address to be specified. A new multi value `spec.loadBalancerIPs` is added;
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: LoadBalancer
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+  ipFamilyPolicy: PreferDualStack
+  loadBalancerIPs:
+  - 10.0.0.1
+  - "1000::1"
+```
+
+While the `spec.loadBalancerIPs` array has become necessary for dual-stack the field has
+a broader use. Kubernetes does not use this field but it is a hint to the cloud-provider
+that may or may not be obeyed. For instance it is a valid case for the user to request
+3 IPv6 load-balancer addresses and no IPv4, or to specify both IPv4 and IPv6 addresses
+in a single-stack service and let the cloud-provider decide which one to use depending
+on the main family.
```

Comment:
Is it assumed that regardless of the number of requested IPs, that the cloud provider will assign at most one IPv4 IP and one IPv6 IP?

Is there really a use case for requesting multiple load balancer IPs like this? How is it that the user knows that 1.2.3.4 and 5.6.7.8 are suitable load balancers, but 9.10.11.12 is not? Why does the cloud provider _not_ know this? It seems like there's secretly some additional information here (eg, topology) and it might be nicer to expose that directly.

This also seems like something that Gateway, etc, might be able to handle better than Service does, in which case maybe it would be better to leave this unsolved in Service?

Reply:
> Is it assumed that regardless of the number of requested IPs, that the cloud provider will assign at most one IPv4 IP and one IPv6 IP?

No, it may differ between cloud-providers I suppose. I have already got queries about multiple load-baclancer addresses for one service so I suppose someone has a use-case (I didn't ask). It was easily solved by defining multiple services, so there are other ways.

I don't know how "any-number" of loadBalancerIPs may be used, but I am sure somebody will have a use-case and IMHO a K8s imposed constraint is unnecessary. This is a user/cloud-provider contract I think K8s should keep out of it.

But I understand the argument that uses may diverge in an undesirable way between cloud-providers if too much freedom is given.

Reply:
> This also seems like something that Gateway, etc, might be able to handle better than Service does, in which case maybe it would be better to leave this unsolved in Service?

I don't understand this. Do you suggest to not add `loadBalancerIPs`? If so, I think this is mostly needed for private-cloud where there is no K8s-aware cloud infra-structure with it's own config setup, but the next step are K8s-unaware routers (GW's). The possibility to specify external addresses (loadBalancerIPs) is a requirement. Now, this _could_ be solved with annotations, but then again almost everything can but it's never pretty.

Reply:
But fir any cloud-providers currently supporting the `loadBalancerIP` field a `loadBalancerIPs` field for dual-stack services must be a wanted feature.

Reply:
> I don't know how "any-number" of loadBalancerIPs may be used, but I am sure somebody will have a use-case and IMHO a K8s imposed constraint is unnecessary. This is a user/cloud-provider contract I think K8s should keep out of it.
> 
> But I understand the argument that uses may diverge in an undesirable way between cloud-providers if too much freedom is given. I am fine with a two-address/ipv4/ipv6 constraint so I will not argue.

I'm not objecting to either way, but I think the semantics need to be well-defined. We don't want `loadBalancerIPs: [ 1.2.3.4, 5.6.7.8 ]` to mean "I want both 1.2.3.4 and 5.6.7.8" on some cloud providers, but "I want 1.2.3.4, or if I can't have that, then I want 5.6.7.8 instead" on others.

> I don't understand this. Do you suggest to not add `loadBalancerIPs`? If so, I think this is mostly needed for private-cloud where there is no K8s-aware cloud infra-structure with it's own config setup, but the next step are K8s-unaware routers (GW's).

OK... it still feels a bit weird to have this essentially only be there for the dumb-private-cloud case. It seems like it would be more k8s-like to match Services to load balancers based on labels or something...

Though I guess it's not really any lower-level than externalIPs

Reply:
The "dumb-private-cloud case" would be all installations without a surrounding K8s aware environment with some cloud-control-manager. Basically all K8s installations not bought or sold and including all "metallb" users. While always a minority in number of installed cluster I don't think those should be dismissed so easily.

But the general use-case must be;
Any installations now using `spec.loadBalancerIP` in some way must have the possibility to use the same function for dual-stack services.

With this in mind, a constraint of two addresses, one of each family is fine.

Reply:
Yes, I wasn't saying we shouldn't support that use case, I was just saying that it would be nicer if there was some way to support that use case well without having to add new fields that are only used for that one use case.

Reply:
I'm not sure how I feel about this loosening.  Why?

Reply:
The `status.loadBalancerIP.ingress` has always been an array. I consider making `loadBalancerIP` singularis was an oversight, it should have been an array long before dual-stack for symmetry. IMHO `loadBalancerIPs` is not a dual-stack thing (but has become important due to dual-stack) and should not have the constraints that applies for other real dual-stack fields such as `clusterIPs`. Not even the "loadBalancerIP=loadBalancerIPs[0]" and not the upgrade/downgrade complexity of those fields. `loadBalancerIPs` should be ignored by K8s precisely as  `loadBalancerIP` today.

But as said, this is my opinion and I am fine with the dual-stack constraints for `loadBalancerIPs`. But I would like this to be decided before I update the PR; should a  `loadBalancerIPs` field be added? Should it have the same constraints as `clusterIPs`?

If multiple  `status.loadBalancerIP.ingress` addresses are specified K8s will setup rules in iptables/ipvs for those addresses. This works already.